### PR TITLE
Libretro Cheat Database Integration

### LINKED ISSUE #663: Increase Bundled Cheat Codes Amount
### What problem does this solve?

The app-bundled `cheats-database.xml` is tiny (only 11 games and 73 total cheats across all systems). It is essentially sample/test data inherited from the original OpenEmu project.

As we increase the coverage to Cheat Codes throughout all the core modules, it would be great to increase the number of bundled cheat codes.

**Acceptance Criteria**
- Cheat Search and Add Cheat features still work
- The few codes from the current database are merged seamless (or removed, if it makes more sense)
- Implementation is done module by module and tested properly, as it was done with Cheat Codes/Search

### Proposed Solution

The main idea is to integrate with the [Libretro Database repository](https://github.com/libretro/libretro-database), since it is very complete and actively maintained.

The best approach seems to be fetching the information on-demand from Libretro on app initialisation + new game added.

1. Fetch the updated .dat file (one per system, ~10-20KB each compressed).
2. Look up existing games MD5 in the .dat → get canonical game name
3. Fetch the .cht file and store locally + store version for updates

- **Pros**: It gets constantly updated, no app size increase, only downloads what is needed.
- **Cons**: It requires to be online when adding new games, adds a bit more latency when starting the app or adding a new game, fails if GitHub/CDN is down.

### Which area does this affect?

Cheat Codes

### Alternatives Considered

Alternatives to this process could be:
- Change download to be triggered by the user (e.g., sync option on Settings)
- Bundle the database in the app and update it with releases

### Apple Silicon / macOS Specific

- [ ] This is specific to Apple Silicon or macOS (not a general OpenEmu request)

### PR DESCRIPTION
**Pull request by leocck**
## What does this PR do?

This PR implements a Browse Online Cheats new feature, based on the existing cheats bundled in the app, with the additional integration with the [Libretro Cheat Database](https://github.com/libretro/libretro-database).

image

image

The architecture was built in a way that we can add other database providers later.

**List of Features:**
- Download / Cache / Update cheat codes from Libretro
- New Browse Online Cheats window with filters and Import button
- Status feedback (working, not working) stored locally
- User notes for a cheat code stored locally
- Cheat normalization / de-duplication

**Important**

**⚠️ Behavior change:** games no longer auto-populate the Cheats menu from the bundled database. The Cheats menu now lists only your own and imported entries. The same bundled cheats are still available — they're served through the new Browse Online Cheats window and added via explicit import.

**Supported Systems**

| System | Core |
|--------|------|
| Atari 2600 | Stella |
| ColecoVision | CrabEmu |
| Famicom Disk System | Nestopia |
| Game Boy | Gambatte |
| Game Boy Advance | mGBA |
| Game Boy Color | Gambatte |
| Game Gear | GenesisPlus |
| Genesis / Mega Drive | GenesisPlus |
| Master System | GenesisPlus |
| Master System | CrabEmu |
| NES / Famicom | FCEU |
| NES / Famicom | Nestopia |
| Nintendo 64 | Mupen64Plus |
| Nintendo DS | DeSmuME |
| Playstation | Mednafen |
| Sega CD | GenesisPlus |
| SG-1000 | GenesisPlus |
| SNES | BSNES |
| SNES | SNES9x |

**Known Issues**

As we are importing cheats from an external database, some systems had unavoidable issues due to inconsistencies on their side and other minor issues.
- **ColecoVision**: CHT files are named as acronims, not the full name described in the DAT files. Implementation is done but will only work after [this PR](https://github.com/libretro/libretro-database/pull/1716) from Libretro-database project is merged. **EDIT**: merged and tested ✅ 
- **SG-1000**: It is not fully supported on Libretro (mismatched directory, mismatched naming conventions). Implementation is done but it will only work after [this PR](https://github.com/libretro/libretro-database/pull/1717) from Libretro-database project is merged. **EDIT**: merged and tested ✅ 
- **Lack of Cheat Grouping**: Libretro imported cheats from GameHacking but they did not carried over the Grouping information. As a result, sometimes the name of the cheats don't give the whole context of what it does.

## What did you test?

- Connection and download of cheat codes from Libretro, for all systems above
- UI normal interactions (Filter, Import, Status Feedback, Remove)
- Edge cases (User cheats with same code, different UI flows, information sync between game and browse window)

## Which cores or systems are affected?

- **DeSmuME**: had to enable cheats and implement a related bugfix
- **GenesisPlus**: had to be extended to provide the serial number of the CD image
- **CrabEmu**: had an issue with cheat codes and I implemented a bugfix

## Did you use AI tools?

Code assistance, adversarial-reviewer agent, task automation

## Linked issues

Fixes #663  

---

## How to test locally

-

---

## PR checklist

- [x] Branched from an up-to-date `main` (ran `git fetch origin && git merge origin/main`)
- [x] Build passes: `./Scripts/verify.sh` (or `xcodebuild -workspace OpenEmu-metal.xcworkspace -scheme OpenEmu -configuration Debug -destination 'platform=macOS,arch=arm64' build`)
- [x] Tested on Apple Silicon (M1 / M2 / M3 / M4 Mac)
- [x] No build logs, binaries, or credentials committed
- [x] Copyright headers preserved on all modified files
- [x] New files (if any) include the BSD 2-Clause license header

### CODE REVIEW

File: OpenEmu/CheatCodeValidator.swift [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    static func isGenesisPARCode(_ code: String) -> Bool {
+        guard code.contains(":") else { return false }
+        let parts = code.split(separator: ":")
+        guard parts.count == 2 else { return false }
+        return parts[0].count == 6 && parts[0].allSatisfy(\.isHexDigit)
+            && parts[1].count >= 1 && parts[1].count <= 4 && parts[1].allSatisfy(\.isHexDigit)
+    }
+
+    /// SMS/GG Action Replay: XXXX-XXXX (2 groups of 4 hex chars)
+    static func isSMSActionReplayCode(_ code: String) -> Bool {
```
Comment (nickybmon) [inline OpenEmu/CheatCodeValidator.swift:221]: This requires two 4-hex groups, but the canonical Sega Action Replay layout is `00AAAA-VV`, which splits 6 and 2 and fails here. CrabEmu handles that form correctly: it strips the dash (`SMSGameCore.m:451`) then reads the 8-char code as prefix + 16-bit address + 8-bit value at `:484`.

So a real code like `00C023-03` gets filtered out by `CheatDatabaseService` before it ever reaches the core that would have applied it. Game Gear and SG-1000 dodge it only because their branch also allows `isSMSGameGenieCode`, which doesn't constrain group length.

Accepting `00AAAA-VV` (2+4 / 2 split) alongside the current shape should cover it.

File: OpenEmu/CheatCodeValidator.swift [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+        default:
+            return true
+        }
+    }
+
+    // MARK: - Format Checks
+
+    /// Raw address:value hex format. Accepts optional exact or max size constraints.
+    static func isRawAddressValue(
```
Comment (nickybmon) [inline OpenEmu/CheatCodeValidator.swift:106]: `split(separator:)` drops empty subsequences, so `1234::56` passes as a valid two-part code. Same pattern in `isGenesisPARCode` (210) and `isSMSGameGenieCode` (139), where `1234--5678` gets through.

`split(separator: ":", omittingEmptySubsequences: false)` would tighten it.

File: OpenEmu/LibretroCheatProvider.swift [hunk 3]
```diff
@@ -0,0 +1,710 @@
...
+                    // Nothing updated yet — return the full cached set as-is
+                    return cached.cheats.map { DatabaseCheat(name: Self.decodingHTMLEntities($0.name), code: $0.code, providerName: name) }
+                } else {
+                    // Some sources updated, this one didn't — keep cached cheats alongside fresh ones
+                    allCheats.append(contentsOf: cached.cheats)
+                }
+            }
+            let cheats = anyUpdated ? dedup(allCheats) : cached.cheats
+            if anyUpdated {
+                saveCachedCheats(LibretroCachedCheatFile(sources: cached.sources, cheats: cheats), md5: md5, systemIdentifier: systemIdentifier)
```
Comment (nickybmon) [inline OpenEmu/LibretroCheatProvider.swift]: The refreshed ETag is dropped here. `downloadCHT` returns `result.etag` on a 200, but this re-persists `cached.sources` unchanged, so the new ETag never lands.

Every launch after the first upstream change sends the stale `If-None-Match`, gets a full 200 body back, and re-parses. The conditional request never converges. Rebuilding `sources` from the download results would fix it.

File: OpenEmu/LibretroCheatProvider.swift [hunk 4]
```diff
@@ -0,0 +1,710 @@
...
+            guard let desc = fields["desc"], !desc.isEmpty else { continue }
+
+            var code = fields["code"] ?? ""
+
+            // If code is empty, try to synthesize from address + value (Format B)
+            // TODO: handle big_endian and memory_search_size for multi-byte systems
+            // TODO: filter by cheat_type — only type 1 (set to value) is usable; types 2-7 need RetroArch's RAM engine
+            if code.isEmpty, let addrStr = fields["address"], let valStr = fields["value"],
+               let addr = UInt(addrStr), let val = UInt(valStr) {
+                code = String(format: "%02X:%02X", addr, val)
```
Comment (nickybmon) [inline OpenEmu/LibretroCheatProvider.swift]: Two things here.

`%02X` is a minimum width, not a fixed one, so an address below 0x1000 emits fewer than 4 digits: `1F:0A`. The NES/FDS validator branch wants exactly 4 address hex chars and SNES wants 6, so these synthesized Format B codes get silently dropped downstream. Padding needs to match the system's expected address width.

Separately, `%X` with a Swift `UInt` is a varargs size mismatch (`UInt` is 64-bit on arm64, `%X` expects 32-bit). `%lX`, or casting to `UInt32`, avoids the UB.

File: OpenEmu/LibretroCheatProvider.swift [hunk 5]
```diff
@@ -0,0 +1,710 @@
...
+    ]
+
+    // Systems where a single system ID maps to multiple Libretro DAT/CHT directories
+    private let systemFallbacks: [String: [String]] = [
+        OESystemIdentifierGB: ["Nintendo - Game Boy", "Nintendo - Game Boy Color"],
+    ]
+
+    // In-memory cache: systemIdentifier → [key → (gameName, libretroSystem)]
+    // Keys are uppercased MD5 hashes and serial numbers
+    private var datCache: [String: [String: (name: String, libretroSystem: String)]] = [:]
```
Comment (nickybmon) [inline OpenEmu/LibretroCheatProvider.swift:86]: `datCache` is mutable state on a singleton (`CheatDatabaseService.shared`), and `cheats(forMD5:…)` is a non-isolated `async` method so it runs off the caller's actor rather than on it.

The per-window `guard !isLoading, !hasLoaded` in the view controller only serializes one window. Two game windows with Browse Online Cheats open both go through the same provider instance and can mutate this dictionary concurrently, which Swift's `Dictionary` doesn't tolerate.

Making the provider an `actor` is the cleanest fix. `OpenEmuCheatProvider`'s `loaded`/`database` have the same exposure, and are additionally touched from the main thread via `supportsSystem` while the cheats menu is built.

File: OpenEmu/LibretroCheatProvider.swift [hunk 6]
```diff
@@ -0,0 +1,710 @@
...
+            guard let underIdx = key.firstIndex(of: "_") else { continue }
+            let prefix = String(key[..<underIdx])
+            let field = String(key[key.index(after: underIdx)...])
+            groups[prefix, default: [:]][field] = value
+        }
+
+        var cheats: [LibretroCachedCheat] = []
+        var seenCodes: Set<String> = []
+
+        for index in 0..<groups.count {
```
Comment (nickybmon) [inline OpenEmu/LibretroCheatProvider.swift:317]: This assumes the prefixes are contiguous `cheat0…cheatN-1`. A gap in the numbering silently truncates the tail: with `cheat0`, `cheat1`, `cheat3` present, `groups.count` is 3, the loop stops at `cheat2`, and `cheat3` is never read.

Iterating the sorted keys of `groups` instead of a count-based range would be immune to both gaps and any stray underscore-containing top-level key.

File: OpenEmu/LibretroCheatProvider.swift [hunk 7]
```diff
@@ -0,0 +1,710 @@
...
+        var ggCodes: [String] = []
+        var i = 0
+        while i < parts.count - 2 {
+            ggCodes.append("\(parts[i])-\(parts[i+1])-\(parts[i+2])")
+            i += 3
+        }
+        return ggCodes.joined(separator: "+")
+    }
+
+    private func normalizePSXCode(_ code: String) -> String {
```
Comment (nickybmon) [inline OpenEmu/LibretroCheatProvider.swift:422]: The O to 0 substitution is applied to `code` before the all-hex guard, and the bail-out path returns the mutated string rather than the original. A code that fails the hex check comes back with its letter Os already replaced by zeros even though normalization was abandoned.

Substituting into a local and only returning it once the guard passes keeps the failure path lossless.

File: OpenEmu/LibretroCheatProvider.swift [hunk 8]
```diff
@@ -0,0 +1,710 @@
...
+            // log.info("Local cache hit for \(md5) (\(cached.sources.map(\.chtFileName).joined(separator: ", ")))")
+            // Try to update each cached source
+            var anyUpdated = false
+            var allCheats: [LibretroCachedCheat] = []
+            for source in cached.sources {
+                if let updated = try await downloadCHT(chtFileName: source.chtFileName, libretroSystem: libretroSystem, systemIdentifier: systemIdentifier, existingETag: source.etag) {
+                    allCheats.append(contentsOf: updated.cheats)
+                    anyUpdated = true
+                } else if !anyUpdated {
+                    // Nothing updated yet — return the full cached set as-is
```
Comment (nickybmon) [inline OpenEmu/LibretroCheatProvider.swift:110]: `parseCHTFile` already ran `decodingHTMLEntities` on the description before storing it in `LibretroCachedCheat` (line 333), so this decodes a second time. Also happens at 118 and 172.

`&amp;amp;quot;` ends up as `"` instead of `&amp;quot;`, which is the exact double-unescape the helper's own comment is guarding against. Decoding once at parse time and mapping straight through here would settle it.

File: OpenEmu/BrowseOnlineCheatsViewController.swift [hunk 9]
```diff
@@ -0,0 +1,1366 @@
...
+                                          comment: "Browse online cheats: disambiguates multiple cheats sharing the same name, e.g. \"99 Magic (2 of 7)\""),
+                cheat.name, index, total
+            )
+            return DatabaseCheat(name: disambiguatedName, code: cheat.code, providerName: cheat.providerName)
+        }
+    }
+
+    /// Feedback is scoped to a core build, so switching cores invalidates both the
+    /// fetched list and the reports shown against it.
+    func resetForCoreChange() {
```
Comment (nickybmon) [inline OpenEmu/BrowseOnlineCheatsViewController.swift:731]: This sets `hasLoaded = false` and calls `fetchOnlineCheats()`, but that guard is `!isLoading, !hasLoaded`. If a fetch is still in flight when the core changes, `isLoading` is still true, the new fetch is skipped, and the in-flight Task then completes and assigns results that were validated against the *previous* core, setting `hasLoaded = true`.

The window ends up showing cheats the new core may not support with no refetch. Holding the Task and cancelling it here before resetting would close it.

File: OpenEmu/BrowseOnlineCheatsViewController.swift [hunk 10]
```diff
@@ -0,0 +1,1366 @@
...
+        gameLabel = makeValueLabel(width: 280)
+        let md5Title = makeTitleLabel(NSLocalizedString("MD5", comment: "Browse online cheats info label"), width: 40)
+        md5Label = makeValueLabel(width: 220)
+
+        let systemTitle = makeTitleLabel(NSLocalizedString("System", comment: "Browse online cheats info label"), width: 60)
+        systemLabel = makeValueLabel(width: 140)
+        let coreTitle = makeTitleLabel(NSLocalizedString("Core", comment: "Browse online cheats info label"), width: 60)
+        coreTitle.alignment = .right
+        coreLabel = makeValueLabel(width: 140)
+        let serialTitle = makeTitleLabel(NSLocalizedString("Version", comment: "Browse online cheats info label"), width: 60)
```
Comment (nickybmon) [inline OpenEmu/BrowseOnlineCheatsViewController.swift:559]: This label reads "Version" but the value at 632 is `document.rom.serial`, so a PSX disc shows `Version: SLUS-00594`. "Serial" would match.

File: OpenEmu/CheatFeedbackService.swift [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+                                               status: status,
+                                               notes: existingNotes,
+                                               updatedAt: Date()))
+
+        save(file, md5: md5, systemIdentifier: systemIdentifier)
+    }
+
+    /// Records a personal note, replacing any previous one for the same code and core build.
+    /// Existing status for that code/build is carried over untouched. `nil`/empty removes the note.
+    func setNotes(_ notes: String?,
```
Comment (nickybmon) [inline OpenEmu/CheatFeedbackService.swift:161]: `setNotes` writes the entry with `existingStatus ?? .unknown`, so saving a note creates a status record for that cheat. `OEGameDocument.swift:2155` then bails on `guard existingStatuses[key] == nil`.

Net effect: adding a note permanently suppresses the "Did this cheat code work for you?" prompt for that cheat on that core build, even though the user never reported anything. Storing notes without materializing a status, or having the prompt check for a real status rather than any record, would separate the two.

File: OpenEmu/OEGameDocument.swift [hunk 12]
```diff
@@ -2053,7 +2111,6 @@ final class OEGameDocument: NSDocument {
```
Comment (nickybmon) [inline OpenEmu/OEGameDocument.swift:2142]: `editCheat` doesn't call `validateCheatCompatibility()` afterwards, unlike `addCheat`, `addCheatFromSearch`, and `addImportedCheat`. The replacement keeps the default `isCompatibleWithCore = true`, so editing a code into a format this core can't handle still shows it as compatible and still offers Enabled, which then does nothing, until the next core switch or relaunch.

File: DeSmuME/src/frontend/cocoa/openemu/NDSGameCore.mm [hunk 13]
```diff
@@ -1239,15 +1240,18 @@ - (void)setCheat:(NSString *)code setType:(NSString *)type setEnabled:(BOOL)enab
...
-		newCheatItem->SetName(NULL); // OpenEmu takes care of this
-		newCheatItem->SetComments(NULL); // OpenEmu does not support cheat item comments
+		newCheatItem->SetName(NULL);
+		newCheatItem->SetComments(NULL);
 		newCheatItem->SetRawCodeString([code cStringUsingEncoding:NSUTF8StringEncoding], true);
 		newCheatItem->SetEnabled((enabled) ? true : false);
 		
 		cocoaCheatItem = [[self cdsCheats] addExistingItem:newCheatItem];
-		
+		if (cocoaCheatItem == nil) {
```
Comment (nickybmon) [inline DeSmuME/src/frontend/cocoa/openemu/NDSGameCore.mm:1322]: `newCheatItem` leaks on this path. `addExistingItem` returns nil when the code is already present (`AddExistingItemNoDuplicate` returns NULL on a duplicate), and the list only takes ownership on success, since `__AddItem` pushes the pointer in directly rather than copying. The `CocoaDSCheatItem` wrapper won't free it either, as `_didAllocateInternalData` is NO when it's constructed from an existing item.

The leak predates this PR, but the early return is the natural place to fix it, and re-importing an already-present cheat becomes a routine action once Browse Online Cheats ships.

```objc
if (cocoaCheatItem == nil) {
    delete newCheatItem;
    return;
}
```

File: OpenEmu/OEGameDocument.swift [hunk 14]
```diff
@@ -1645,23 +1661,66 @@ final class OEGameDocument: NSDocument {
...
             setCheat(cheat)
         }
         cheats.append(cheat)
         saveUserCheats()
+        validateCheatCompatibility()
+    }
+
+    /// `cheatSource` (the provider name) is what marks this as Browse Online Cheats-imported,
+    /// distinguishing it from cheats added manually or via Cheat Search.
+    func addImportedCheat(code: String, name: String, providerName: String) {
```
Comment (nickybmon) [inline OpenEmu/OEGameDocument.swift:1683]: Everything imported comes through as `OECheatTypeGameShark` here, and `DatabaseCheat` doesn't carry a type at all, so I'm guessing this was a simplification to get the provider interface landed.

Where it bites: the validator allows raw `6hex:2hex` SNES codes, but BSNES only strips the colon for Action Replay or Raw (`BSNESGameCore.mm:384`). So `7E0019:63` reaches `OEBSNESCheatDecodeSNES` with the colon still on it, matches none of its four shapes, and comes back false. The cheat sits there enabled doing nothing. Tagged Raw it strips to `7e001963` and decodes fine as an 8-char PAR code.

It also drops the per-cheat type the bundled XML was carrying through `Cheats.swift`. Adding a type to `DatabaseCheat` looks like the answer, but it touches both providers and you may have had a reason to keep that struct thin, so worth settling before building.

File: OpenEmu/OEGameDocument.swift [hunk 15]
```diff
@@ -1645,23 +1661,66 @@ final class OEGameDocument: NSDocument {
...
+        if cheat.isEnabled {
+            gameCoreManager?.setCheat(cheat.code, withType: cheat.type, enabled: false)
+        }
+        cheats.remove(at: index)
+        saveUserCheats()
+        promptCheatRemovalFeedback(code: cheat.code)
     }
     
     /// In order to load cheats, we need the core plugin and the ROM to be set.
     private func loadCheats() {
```
Comment (nickybmon) [inline OpenEmu/OEGameDocument.swift:1712]: With the `Cheats(md5Hash:)` path gone there are no callers left, so the Cheats menu is user and imported entries only now. Nothing's actually lost since `OpenEmuCheatProvider` serves the same XML through the new window, and I assume the intent is moving bundled cheats behind an explicit import rather than auto-populating them.

Flagging it mainly because it changes what existing users see: a game that used to list cheats now opens with an empty menu until they go and import. If that's the direction, it wants a release note and a line in the description.

**Review by nickybmon — CHANGES_REQUESTED** [review]
Left comments in line and on the PR!

File: OpenEmu/CheatCodeValidator.swift (hunk 1 above)
Comment (leocck) [reply OpenEmu/CheatCodeValidator.swift:221]: I don't think this is correct. Libretro .cht files for every SMS/GG Action Replay code is written XXXX-XXXX (4 hex + dash + 4 hex), not 00AAAA-VV (6+2). Same format is being used in the embedded cheat xml file.

I noticed there was a comment on CrabEmu's core with the wrong cheat format documented, so I fixed it to avoid the same confusion in the future.

File: OpenEmu/CheatCodeValidator.swift (hunk 2 above)
Comment (leocck) [reply OpenEmu/CheatCodeValidator.swift:106]: Validation added, even though there is no cheat code like these in the databases. A code like this would be definitely a typo.

File: OpenEmu/LibretroCheatProvider.swift (hunk 3 above)
Comment (leocck) [reply OpenEmu/LibretroCheatProvider.swift]: Fixed

File: OpenEmu/LibretroCheatProvider.swift (hunk 4 above)
Comment (leocck) [reply OpenEmu/LibretroCheatProvider.swift]: both fixed

File: OpenEmu/LibretroCheatProvider.swift (hunk 5 above)
Comment (leocck) [reply OpenEmu/LibretroCheatProvider.swift:86]: That was a tricky one. I fixed the race conditions with locks rather than converting to actors, since supportsSystem has to stay synchronous for AppKit cheats-menu building (GameControlsBar → supportsOnlineCheats)

File: OpenEmu/CheatFeedbackService.swift (hunk 11 above)
Comment (leocck) [reply OpenEmu/CheatFeedbackService.swift:161]: Well spot, and this made me realise I could extend the Notes dialog to add the feedback options as a convenience but optional.

File: OpenEmu/OEGameDocument.swift (hunk 12 above)
Comment (leocck) [reply OpenEmu/OEGameDocument.swift:2142]: Nice catch, fixed

File: DeSmuME/src/frontend/cocoa/openemu/NDSGameCore.mm (hunk 13 above)
Comment (leocck) [reply DeSmuME/src/frontend/cocoa/openemu/NDSGameCore.mm:1322]: Fixed. The early return now `delete newCheatItem;` before returning, since `addExistingItem:` doesn't take ownership on the duplicate path.

File: OpenEmu/OEGameDocument.swift (hunk 14 above)
Comment (leocck) [reply OpenEmu/OEGameDocument.swift:1683]: Yes, the dropping of the code type was intentional because it was not reliable and not adding much to the cheat validations. 

This was a bug though and I fixed like the other core systems do: derived the type from the code shape at the import boundary.

File: OpenEmu/OEGameDocument.swift (hunk 15 above)
Comment (leocck) [reply OpenEmu/OEGameDocument.swift:1712]: Correct on both counts. It was intentional, bundled cheats now come through Browse Online Cheats via explicit import. Added the behavior-change callout to the PR description. We should mention this in the release note for the shipping version.

**Review by nickybmon — APPROVED** [review]
